src/crimson/common/errorator.h

```diff
@@ -353,7 +353,8 @@ class maybe_handle_error_t {
         // execute for the sake a side effects. Typically this boils down
         // to throwing an exception by the handler.
         std::invoke(std::forward<ErrorVisitorT>(errfunc),
-                    ErrorT::error_t::from_exception_ptr(std::move(ep)));
+                    ErrorT::error_t::from_exception_ptr(ep));
+        result = FuturatorT::make_exception_future(std::move(ep));
```

Comment:
> Not sure whether "passing a void-return handler would pass on the error, instead of consuming it" is reasonable.
> 
> IIUC if there is a handler to handle a specfic error, the error should be consumed unless the handler itself returns the error again.

Yeah, but the problem is that there seems to be no way for us to know what value should be returned for a specific future type unless that future type's value is default-constructible. I tried to disallow void-return handlers, but we need the `discard_all` and `assert_all` interfaces. So I came to this PR's implementation.

On the other hand, how about we mandate that the future value must be default-constructible if a void-returning error handler is needed?

Reply:
Seems to me it is user's fault if the error-handler implementation returns void but the returning valid future expects non-void results.

It should end up with compile failure. And probably we can make the error message more friendly with some static asserts?

Reply:
`discard_all` seem to have no users, and it doesn't seem to be able/reasonable to to construct any valid results.

For `assert_all`, is it possible to make it to return a special label (or check the type at runtime?) so it can be handled differently?

cc @rzarzynski @athanatos

Reply:
> `discard_all` seem to have no users, and it doesn't seem to be able/reasonable to to construct any valid results.

`discard_all` is used in seastore unittests. I think `discard_all` needs to return a valid result so that the error is "discarded" and the continuation chain can continue, am I missing something?

> For `assert_all`, is it possible to make it to return a special label (or check the type at runtime?) so it can be handled differently?

I'll give it a try.

Reply:
> discard_all is used in seastore unittests. I think discard_all needs to return a valid result so that the error is "discarded" and the continuation chain can continue, am I missing something?

Seems the only user is originated from the commit https://github.com/ceph/ceph/commit/e936bcdfc39d2719c93d12d45604d6076a0afb03, should `discard_all` to be converted to an `assert_all` to indicated that any error is not expected during the test?

Reply:
Ah, yes.

Reply:
All done. Please take a look again, thanks.